File: section_3.1/data_importer.R

getNonVarCols <- function(data) {
  var_columns <- apply(data, 2, var)
  resCol <- names(var_columns[var_columns == 0.0])
  return(resCol)
}

get.data <- function(data_dir, data_source, dataset) {
  dataset_dir <- file.path(data_dir, dataset)
  if (data_source == "chipseq") {
    inputFileName <- file.path(dataset_dir, 'inputs.csv')
    labelFileName <- file.path(dataset_dir, 'outputs.csv')
    foldsFileName <- file.path(dataset_dir, 'cv/equal_labels/folds.csv')
    inputs <- read.table(inputFileName, sep=",", header=TRUE, stringsAsFactors=FALSE, row.names=1)
    labels <- read.table(labelFileName, sep=",", header=TRUE, stringsAsFactors=FALSE, row.names=1)
    colnames(labels) <- c('min.log.penalty', 'max.log.penalty')
    folds <- read.table(foldsFileName, sep=",", header=TRUE, stringsAsFactors=FALSE, row.names=1)

    rownamesInput <- rownames(inputs)
    inputs <- do.call(data.frame,lapply(inputs, function(x) replace(x, is.infinite(x), NA)))
    naColumns <- colnames(inputs)[colSums(is.na(inputs)) > 0]
    noVarCol <- getNonVarCols(inputs)
    removeCols <- c(naColumns,noVarCol)
    inputs <- inputs[, !(colnames(inputs) %in% removeCols)]
    rownames(inputs) <- rownamesInput

    X <- inputs
    y <- labels
  } else if (data_source == "simulated") {
    X_path <- file.path(dataset_dir, 'features.csv')
    y_path <- file.path(dataset_dir, 'targets.csv')
    folds_path <- file.path(dataset_dir, 'folds.csv')
    X <- read.table(X_path, sep=",", header=TRUE, stringsAsFactors=FALSE)
    y <- read.table(y_path, sep=",", header=TRUE, stringsAsFactors=FALSE)
    folds <- read.table(folds_path, sep=",", header=TRUE, stringsAsFactors=FALSE)
  } else {
    stop("data_source must be one of {chipseq,simulated}")
  }
  result <- list()
  result$X <- X
  result$y <- y
  result$folds <- folds
  return(result)
}

File: section_3.1/data_importer.py
import numpy as np
import pandas as pd
import os

def data_import(current_dir, data_source, data_name):
    data_dir = os.path.join(current_dir, os.path.pardir, 'data', data_name)
    assert os.path.isdir(data_dir)
    if data_source == 'simulated':
        inputFileName = os.path.join(data_dir, 'features.csv')
        labelFileName = os.path.join(data_dir, 'targets.csv')
        foldsFileName = os.path.join(data_dir, 'folds.csv')
        inputs = pd.read_csv(inputFileName).astype(np.double)
        labels = pd.read_csv(labelFileName).astype(np.double)
        folds = pd.read_csv(foldsFileName)
        # pre-processing: change label scale
        labels['min.lambda'] = labels['min.log.penalty'].apply(lambda x: np.exp(float(x)))
        labels['max.lambda'] = labels[' max.log.penalty'].apply(lambda x: np.exp(float(x)))
    elif data_source == 'chipseq':
        inputFileName = os.path.join(data_dir, 'inputs.csv')
        labelFileName = os.path.join(data_dir, 'outputs.csv')
        foldsFileName = os.path.join(data_dir, 'cv/equal_labels/folds.csv')
        inputs = pd.read_csv(inputFileName, index_col='sequenceID').astype(np.double)
        labels = pd.read_csv(labelFileName, index_col='sequenceID').astype(np.double)
        folds = pd.read_csv(foldsFileName, index_col='sequenceID')

        # pre-processing: remove all data columns with INFs and NaNs
        inputs.replace([-float('inf'), float('inf')], np.nan, inplace=True)
        missingCols = inputs.isnull().sum()
        missingCols = list(missingCols[missingCols > 0].index)
        print(f'missingCols = {missingCols}')
        inputs.drop(missingCols, axis=1, inplace=True)
        # pre-processing: remove all data columns with zero variance
        varCols = inputs.apply(lambda x: np.var(x))
        zeroVarCols = list(varCols[varCols == 0].index)
        print(f'zeroVarCols = {zeroVarCols}')
        inputs.drop(zeroVarCols, axis=1, inplace=True)
        # pre-processing: change label scale
        labels['min.lambda'] = labels['min.log.lambda'].apply(lambda x: np.exp(x))
        labels['max.lambda'] = labels['max.log.lambda'].apply(lambda x: np.exp(x))
    else:
        raise ValueError(f'Unknown data source {data_source}')

    return inputs, labels, folds

